File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.ts
function smallestPalindrome(s: string): string {
    const n = s.length;
    const freq = new Array(26).fill(0);
    const arr = s.split('');

    // Count first half
    for (let i = 0; i < n >> 1; i++) {
        const idx = (s.charCodeAt(i) & 31) - 1;
        freq[idx]++;
    }

    let j = 0;
    for (let i = 0; i < 26; i++) {
        while (freq[i] > 0) {
            arr[j] = String.fromCharCode(97 + i);
            arr[n - 1 - j] = String.fromCharCode(97 + i);
            j++;
            freq[i]--;
        }
    }

    return arr.join('');
}

File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.rs
impl Solution {
    pub fn smallest_palindrome(s: String) -> String {
        let n = s.len();
        let mut freq = [0; 26];
        let bytes = s.as_bytes();

        // Count first half
        for i in 0..n/2 {
            let idx = (bytes[i] & 31) as usize - 1;
            freq[idx] += 1;
        }

        let mut arr: Vec<u8> = bytes.to_vec();
        let mut j = 0;
        for i in 0..26 {
            while freq[i] > 0 {
                arr[j] = (b'a' + i as u8);
                arr[n - 1 - j] = (b'a' + i as u8);
                j += 1;
                freq[i] -= 1;
            }
        }

        String::from_utf8(arr).unwrap()
    }
}